core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/jndi/JndiSupport.scala

```diff
@@ -34,8 +36,31 @@ trait JndiSupport extends LazyLogging {
     t.get
   }
 
+  protected def withSearchResults[T](ctx: DirContext, name: String, matchingAttributes: Attributes, attributesToReturn: Array[String] = Array.empty)(op: Iterator[SearchResult] => T): T = {
+    if (attributesToReturn.isEmpty) {
+      handleSearchResults(op, ctx.search(name, matchingAttributes))
+    } else {
+      handleSearchResults(op, ctx.search(name, matchingAttributes, attributesToReturn))
+    }
+  }
+
+  protected def withSearchResults[T](ctx: DirContext, name: String, filter: String, cons: SearchControls)(op: Iterator[SearchResult] => T): T = {
+    handleSearchResults(op, ctx.search(name, filter, cons))
+  }
+
+  private def handleSearchResults[T](op: (Iterator[SearchResult]) => T, results: NamingEnumeration[SearchResult]) = {
+    // need to copy the results out of the jndi enum otherwise scala is fancy and continues to use the
+    // enum behind the scenes which is a problem once the enum is closed :(
+    val b = new mutable.ListBuffer[SearchResult]
```

Comment:
Thank you for explaining the need to copy the results.

Why use `mutable.ListBuffer`? Then why pass `.toIterator` to `op`? Many ops then `.toSeq` the results and then match on `Seq()` and `Seq(something)`. Does it make sense to shortcut some of the conversions (even just for readability) and then match on `Nil` and `something :: Nil` thereby hopefully only caring about the specific collection types when it truly matters?

I'm mainly interested in the readability of callers of `withSearchResults`. As it is now, I look at `.toSeq` and wonder what the type is that it needs to be changed which then leads me to this function and few answers. If callers could be written more generally, I wouldn't even have to wonder and end up here.

(This is as much of a question as it is a suggestion because I'm not sure what I'm proposing would actually work.)

Reply:
generally I was going for a drop-in replacement. The asScala in the old code resulted in an Iterator so that informed the current type. Using the ListBuffer came later when I ran into problems accessing the collection after the underlying ldap enum was closed. ListBuffer seemed better than the other option-ArrayBuffer. I have been told that Seqs are better than Lists. Pattern matching on Seq is just as valid as Lists.